editor/src/core/workers/parser-printer/uid-fix.ts

```diff
@@ -1,284 +1,569 @@
+  mappings: Array<{ oldUID: string; newUID: string }>
+}
 
 export function fixParseSuccessUIDs(
   oldParsed: ParseSuccess | null,
   newParsed: ParsedTextFile,
   alreadyExistingUIDs: Set<string>,
 ): ParsedTextFile {
-  if (oldParsed == null || !isParseSuccess(newParsed)) {
+  if (oldParsed == null || newParsed.type !== 'PARSE_SUCCESS') {
     // we won't try to fix parse failures
     return newParsed
   }
 
-  let newToOldUidMapping: {
-    [newUID: string]: {
-      oldUID: string
-      newUID: string
-      pathToModify: StaticElementPathPart
-      oldPathToRestore: StaticElementPathPart
-    }
-  } = {}
+  // This gets passed through all the fixing functions, so
+  // that the changes can be recorded.
+  const fixUIDsState: FixUIDsState = {
+    mutableAllNewUIDs: alreadyExistingUIDs,
+    mappings: [],
+  }
 
-  const numberOfElementsMatch = zipTopLevelElements(
+  // Fix the UIDs in the content.
+  const fixedTopLevelElements = fixTopLevelElementsUIDs(
     oldParsed.topLevelElements,
     newParsed.topLevelElements,
-    (
-      oldUID: string,
-      newUID: string,
-      oldPathToRestore: StaticElementPathPart,
-      newPath: StaticElementPathPart,
-    ) => {
-      if (oldUID !== newUID && !alreadyExistingUIDs.has(oldUID)) {
-        // we have a UID mismatch and have confirmed that UID doesn't exist elsewhere
-        newToOldUidMapping[newUID] = {
-          oldUID: oldUID,
-          newUID: newUID,
-          pathToModify: newPath,
-          oldPathToRestore: oldPathToRestore,
-        }
-      }
-    },
+    fixUIDsState,
   )
-
-  const newToOldUidMappingArray = Object.values(newToOldUidMapping)
-
-  if (newToOldUidMappingArray.length === 1 || numberOfElementsMatch) {
-    // we found a single UID mismatch, or the total number of elements has stayed the same,
-    // which means there's a very good chance that it was an update to existing elements,
-    // or a single new element was inserted
-    let workingComponents = getComponentsFromTopLevelElements(newParsed.topLevelElements)
-
-    // We need to sort the array first, as the pathToModify will be based on the new UIDs up to the leaf,
-    // so we must update the deepest elements first
-    newToOldUidMappingArray.sort((l, r) => r.pathToModify.length - l.pathToModify.length)
-
-    newToOldUidMappingArray.forEach((mapping) => {
-      const oldPathAlreadyExistingElement = findJSXElementChildAtPath(
-        workingComponents,
-        EP.elementPath([mapping.oldPathToRestore]),
-      )
-
-      if (oldPathAlreadyExistingElement == null) {
-        workingComponents = transformJSXComponentAtElementPath(
-          workingComponents,
-          mapping.pathToModify,
-          (element) => {
-            return setUtopiaID(element, mapping.oldUID)
-          },
+  const fixedCombinedTopLevelArbitraryBlock =
+    oldParsed.combinedTopLevelArbitraryBlock == null ||
+    newParsed.combinedTopLevelArbitraryBlock == null
+      ? newParsed.combinedTopLevelArbitraryBlock
```

Comment:
I think that for clarity this could be explicitly be `null`

```ts
const fixedCombinedTopLevelArbitraryBlock =
  oldParsed.combinedTopLevelArbitraryBlock == null ||
  newParsed.combinedTopLevelArbitraryBlock == null
    ? null
    : fixArbitraryJSBlockUIDs(
        oldParsed.combinedTopLevelArbitraryBlock,
        newParsed.combinedTopLevelArbitraryBlock,
        fixUIDsState,
      )
```

Reply:
uh, that's true 🤦